### PR TITLE
flush gpu queue

### PR DESCRIPTION
**Pull request by PhilipVinc**
It's not nice, and it will flush twice the cpu if we only have cpus, but jax does not have a stable api to get a list of all devices.
I think it's ok for now?

### CODE REVIEW

**Comment by dionhaefner** [comment]
Yeah I thought a bit about this one. This is somewhat icky, because it will flush the GPU just because you own one, even when using `JAX_PLATFORM_NAME=cpu` and JAX never touches the GPU. I guess this wouldn't be a problem in 99% of cases, but with the greedy JAX preallocation you could get some nasty crashes when JAX tries to preallocate the GPU for the first time during shutdown.

My idea was to set a global flag in our GPU functions. Then we can flush GPUs only when users actually use `mpi4jax` on GPU. Does that make sense?

**Comment by PhilipVinc** [comment]
I had also thought about that road...

It does add a tiny bit of side-effect to our functions, which I dislike. For example if jax will ever support serialising compiled functions our behaviour would probably break.

I'll admit that's far off, however. So maybe we can do as you say, possibly with a decorator on all mpi functions so that the code can easily be refactored?

File: mpi4jax/flush.py [hunk 1]
```diff
@@ -7,3 +7,8 @@ def flush():
     for device in jax.devices("cpu"):
```
Comment (dionhaefner) [inline mpi4jax/flush.py:7]: Can't this just be 

```python
devices = [*jax.devices("cpu"), *jax.devices("gpu")]
for device in devices:
    ...
```

?

**Comment by dionhaefner** [comment]
Sure, let's do that.

File: mpi4jax/flush.py (hunk 1 above)
Comment (PhilipVinc) [reply mpi4jax/flush.py:7]: `jax.devices("gpu")` fails if you don't have a gpu.

You should probe what are the available devices, but that is an implementation detail subject to change in jaxlib and you also get extra devices (like interpreters...) that don't mean anything.

Comment (dionhaefner) [reply mpi4jax/flush.py:7]: I think I would prefer

```python
devices = jax.devices("cpu")

try:
    devices.extend(jax.devices("gpu"))
except RuntimeError:
    pass
```

then. Also prevents the double flush on CPU.

**Comment by dionhaefner** [comment]
I can take care of the decorator if you're not already on it, I had some ideas for that.

**Comment by PhilipVinc** [comment]
If you want to do it, go ahead. Setting up the CUDA docker image and integrating with SLURM for CI is revealing quite complex...

I will merge this for now, anyway, if that's ok?

**Comment by dionhaefner** [comment]
Sure.

**Comment by dionhaefner** [comment]
Why Docker? Are you doing this on Github Actions or somewhere else?

**Comment by PhilipVinc** [comment]
No, I am running some tests on my workstation at EPFL and eventually it will be running on the EPFL cluster. Docker is what IT requires. Build cannot run as a standard user.

Runs will be triggered by a comment by us, something like `@something run GPU` .

**Pull request merged**